crates/vm/levm/src/precompiles.rs

```diff
@@ -501,12 +501,186 @@ pub fn ecmul(
     }
 }
 
-fn ecpairing(
-    _calldata: &Bytes,
-    _gas_for_call: u64,
-    _consumed_gas: &mut u64,
+pub fn ecpairing(
+    calldata: &Bytes,
+    gas_for_call: u64,
+    consumed_gas: &mut u64,
 ) -> Result<Bytes, VMError> {
-    Ok(Bytes::new())
+    // The input must always be a multiple of 192 (6 32-byte values)
+    if calldata.len() % 192 != 0 {
+        return Err(VMError::PrecompileError(PrecompileError::ParsingInputError));
+    }
+
+    let groups_number = calldata.len() / 192;
+
+    // Consume gas
+    let gas_cost = gas_cost::ecpairing(groups_number)?;
+    increase_precompile_consumed_gas(gas_for_call, gas_cost, consumed_gas)?;
+
+    let mut mul: FieldElement<Degree12ExtensionField> = QuadraticExtensionFieldElement::one();
+    for group_number in 0..groups_number {
+        // Define the group indexes and slice calldata to get the group data
+        let group_start = group_number
+            .checked_mul(192)
+            .ok_or(InternalError::ArithmeticOperationOverflow)?;
+        let next_group_start = (group_number
+            .checked_add(1)
+            .ok_or(InternalError::ArithmeticOperationOverflow)?)
+        .checked_mul(192)
+        .ok_or(InternalError::ArithmeticOperationOverflow)?;
+
+        let group_data = calldata
+            .get(group_start..next_group_start)
+            .ok_or(InternalError::SlicingError)?;
+
+        let first_point_x = group_data.get(..32).ok_or(InternalError::SlicingError)?;
+        let first_point_y = group_data.get(32..64).ok_or(InternalError::SlicingError)?;
+
+        // Infinite is defined by (0,0). Any other zero-combination is invalid
+        if (U256::from_big_endian(first_point_x) == U256::zero()
+            && U256::from_big_endian(first_point_y) != U256::zero())
+            || (U256::from_big_endian(first_point_x) != U256::zero()
+                && U256::from_big_endian(first_point_y) == U256::zero())
+        {
+            return Err(VMError::PrecompileError(PrecompileError::DefaultError));
+        }
+
+        let first_point_y = BN254FieldElement::from_bytes_be(first_point_y)
+            .map_err(|_| PrecompileError::DefaultError)?;
+        let first_point_x = BN254FieldElement::from_bytes_be(first_point_x)
+            .map_err(|_| PrecompileError::DefaultError)?;
+
+        let second_point_x_first_part =
+            group_data.get(96..128).ok_or(InternalError::SlicingError)?;
+        let second_point_x_second_part =
+            group_data.get(64..96).ok_or(InternalError::SlicingError)?;
```

Comment:
This got me confused. From 64..96 we have x2 and from 96..128 we have y2 according to evm codes. Why does this look different?

Reply:
I gave it that name because of the place they occupy in the vector, maybe it is not very clear. I could give it another name, but I focused on the fact that, after all, it is a component of what will be the second point.